tutorial/ndarrays.md

```diff
@@ -0,0 +1,328 @@
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+     |error [semantic]: ex.py [4]| Incompatible redefinition (|a| real(10, 20) <-> real(10,))
+    ```
+
+    This limitation is due to the way Fortran allocatable can't change the rank after declaration.
+
+-   Can not assign ndarrays that own their data one another.
+
+    ```Python
+    import numpy as np
+
+    a = np.array([1, 2, 3, 4, 5])
+    b = np.array([1, 2, 3])
+    a = b
+    ```
+
+    _OUTPUT_ :
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+     |error [semantic]: ex.py [5]| Arrays which own their data cannot become views on other arrays (a)
+    ```
+
+    This limitation is due to the fact that the ndarray **a** will have to go from a data owner to a pointer to the **b** ndarray data.
+
+    _NOTE_: this limitation does not include reassigning using new data with respecting the previous rule.
+
+    -   Python example:
+
+        ```Python
+        import numpy as np
+        a = np.ones(20)
+        #(some code...)
+        a = np.ones(10)
+        ```
+
+    -   C equivalent:
+
+        ```C
+        #include <ndarrays.h>
+        #include <stdlib.h>
+        int main()
+        {
+            t_ndarray a;
+            a = array_create(1, (int32_t[]){20}, nd_double);
+            array_fill((double)1.0, a);
+            /*(some code...)*/
+            if (a.shape != NULL)
+            {
+                free_array(a); //this line handles the redefinition of the array
+            }
+            a = array_create(1, (int32_t[]){10}, nd_double);
+            array_fill((double)1.0, a);
+            free_array(a); //garbage collection at the end of the program
+            return 0;
+        }
+        ```
+
+    -   Fortran equivalent:
+
+        ```Fortran
+        program prog_ex
+
+        use, intrinsic :: ISO_C_BINDING
+
+        implicit none
+
+        real(C_DOUBLE), allocatable :: a(:)
+
+        allocate(a(0:20_C_INT64_T - 1_C_INT64_T))
+        a = 1.0_C_DOUBLE
+        !(some code...)
+        if (allocated(a)) then
+        if (any(size(a) /= [10_C_INT64_T])) then
+          deallocate(a)
+          allocate(a(0:10_C_INT64_T - 1_C_INT64_T))
+        end if
+        else
+        allocate(a(0:10_C_INT64_T - 1_C_INT64_T))
+        end if
+        a = 1.0_C_DOUBLE
+
+        end program prog_ex
+        ```
+
+-   Can not reassign to a ndarray that has another pointer accessing his data.
+
+    ```Python
+    import numpy as np
+
+    a = np.ones(10)
+    b = a[:5]
+    #(some code...)
+    a = np.zeros(20)
+    ```
+
+    _OUTPUT_ :
+
+    ```Shell
+    ERROR at annotation (semantic) stage
+    pyccel:
+      |error [semantic]: ex.py [6]| Attempt to reallocate an array which is being used by another variable (a)
+    ```
+
+    This limitation is set since we need to free the previous data when are trying to reallocate the ndarray, which in this case will cause the data where the view **b** point to became inaccessible.
```

Comment:
English improvement:
```
This limitation is set since we need to free the previous data when we reallocate the ndarray. In this case, this will cause the data pointed to by the view **b** to became inaccessible
```